src/ruby/qps/proxy-worker.rb

```diff
@@ -41,32 +41,54 @@ def setup(config)
     @histmax = config.histogram_params.max_possible
     @histogram = Histogram.new(@histres, @histmax)
     @start_time = Time.now
-    # TODO(vjpai): Support multiple client channels by spawning off a PHP client per channel
-    if @use_c_ext
-      puts "Use protobuf c extension"
-      command = "php -d extension=" + File.expand_path(File.dirname(__FILE__)) + "/../../php/tests/qps/vendor/google/protobuf/php/ext/google/protobuf/modules/protobuf.so " + "-d extension=" + File.expand_path(File.dirname(__FILE__)) + "/../../php/ext/grpc/modules/grpc.so " + File.expand_path(File.dirname(__FILE__)) + "/../../php/tests/qps/client.php " + @mytarget
-    else
-      puts "Use protobuf php extension"
-      command = "php -d extension=" + File.expand_path(File.dirname(__FILE__)) + "/../../php/ext/grpc/modules/grpc.so " + File.expand_path(File.dirname(__FILE__)) + "/../../php/tests/qps/client.php " + @mytarget
-    end	
-    puts "Starting command: " + command
-    @php_pid = spawn(command)
+    @php_pid = Array.new(@config.client_channels)
+    (0..@config.client_channels-1).each do |chan|
+      Thread.new {
+        if @use_c_ext
+          puts "Use protobuf c extension"
+          command = "php -d extension=" + File.expand_path(File.dirname(__FILE__)) +
+            "/../../php/tests/qps/vendor/google/protobuf/php/ext/google/protobuf/modules/protobuf.so " +
+            "-d extension=" + File.expand_path(File.dirname(__FILE__)) + "/../../php/ext/grpc/modules/grpc.so " +
+            File.expand_path(File.dirname(__FILE__)) + "/../../php/tests/qps/client.php " + @mytarget + " #{chan%@config.server_targets.length}"
+        else
+          puts "Use protobuf php extension"
+          command = "php -d extension=" + File.expand_path(File.dirname(__FILE__)) + "/../../php/ext/grpc/modules/grpc.so " +
+            File.expand_path(File.dirname(__FILE__)) + "/../../php/tests/qps/client.php " + @mytarget + " #{chan%@config.server_targets.length}"
+        end
+        puts "[ruby proxy] Starting #{chan}th php-client command use c protobuf #{@use_c_ext}: " + command
+        @php_pid[chan] = spawn(command)
+        while true
+          sleep
+        end
+      }
+    end
   end
   def stop
-    Process.kill("TERM", @php_pid)
-    Process.wait(@php_pid)
+    (0..@config.client_channels-1).each do |chan|
+      Process.kill("TERM", @php_pid[chan])
+      Process.wait(@php_pid[chan])
+    end
   end
   def get_config(_args, _call)
-    puts "Answering get_config"
     @config
   end
   def report_time(call)
-    puts "Starting a time reporting stream"
     call.each_remote_read do |lat|
       @histogram.add((lat.latency)*1e9)
     end
     Grpc::Testing::Void.new
   end
+  def report_hist(call)
```

Comment:
Just come in mind. I am investigating multiple php client call it will cause race condition for `@histogram` or not.

Reply:
Not.

Reply:
I actually don't think that's correct. You may have gotten lucky and not had a race when you ran it, but the Ruby server implementation spawns a new thread for a streaming RPC (and you should have @apolcyn review this also given the amount of Ruby changes you've made). So, technically there could be races between the different clients as they operate in this function. To avoid a potential race, just make the histogram have a lock and synchronize on it in the revised `add_contents` function that I've suggested above.

Reply:
I was testing it by atomicly adding ad test_count and compare it with the @histogram.count. I may get lucky. I'll synchronize histogram when merging them. 
```
@lock.synchronize do
  test_count += 10
end
```

Reply:
Done. I add a "Mutex" inside histogram class. I think it can avoid race when multiple client call "report_hist". @apolcyn . Thank you for helping me review the pull request!